cpp/include/cudf/detail/utilities/cast_functor.cuh

```diff
@@ -31,6 +31,18 @@
 namespace cudf {
 namespace detail {
 
+/**
+ * @brief Functor that casts a primitive input value to a specified type
+ */
+template <typename T>
+struct cast_fn {
+  template <typename U>
+  CUDF_HOST_DEVICE constexpr T operator()(U val) const
+  {
+    return static_cast<T>(val);
+  }
```

Comment:
We could also add a direct forwarding of `T` so that we do not need to copy
```suggestion
  CUDF_HOST_DEVICE constexpr T operator()(U val) const
  {
    return static_cast<T>(val);
  }
  
  CUDF_HOST_DEVICE constexpr T&& operator()(T&& val) const noexcept
  {
    return cuda::std::forward<T>(val);
  }
```